058-kraft-upgrades-and-downgrades.md

```diff
@@ -0,0 +1,221 @@
+   Since the user triggers the downgrade by manual action (editing the Kafka CR), it is expected that this error will be caught on early and the user will fix the problem or revert the downgrade.
+2. Next, the operator will roll the Kafka pods to use the new container with the desired version.
+
+_Note:_
+_This corresponds to what Strimzi supports today for ZooKeeper-based clusters._
+_Today, the users are responsible for downgrading the `inter.broker.protocol.version` before the Kafka downgrade by updating their `Kafka` CR._
+_Similarly, even today they have to downgrade Strimzi versions step by step in the same way as described in this section._
+
+#### (Stretch goal) Downgrade of Kafka after Strimzi Cluster Operator downgrade
+
+As a _stretch_ goal, we should also support Kafka downgrades right after the Strimzi downgrade (i.e. jumping multiple Strimzi and Kafka versions during downgrade).
+Implementing this has additional technical complications because it needs to understand Kafka versions that did not exist at the time the software was released.
+While this feature might be useful in some situations, it does not need to be implemented to consider our KRaft support _production-ready_.
+So it might be implemented only later after the other parts.
+
+Even in this downgrade scenario, the user has to make sure the metadata version used by Strimzi is valid even for the Kafka version that we should downgrade to.
+After the Cluster Operator is downgraded, the operator will validate the Kafka version and the metadata version.
+And if the validation passes, it will automatically roll all Kafka pods to use the Kafka version we are downgrading to.
+If the validation fails, it will be up to the user to address the issue.
+This might require moving back to the original Strimzi and Kafka version.
+
+#### Example YAML files
+
+Given the current limitations when downgrading the metadata versions, it seems reasonable for users to prefer managing the metadata manually rather then to have them changed automatically.
+So we should encourage users to set the `Kafka.spec.kafka.metadataVersion` by having it set in our KRaft example YAML files.
+It will be set there to the metadata version corresponding to the Kafka version the example files use.
+This would also correspond to the current ZooKeeper-based examples that have the `inter.broker.protocol.version` set.
+
+#### Risks
+
+Kafka provides APIs to manage the `metadata.version`.
+Strimzi has no way to block the users from using these APIs.
+If the users decide to manipulate the `matadata.version` them self, this can lead to unexpected issues.
+This risk is deemed acceptable, because we expect that:
+* Users would prefer to have the `metadata.version` managed through Strimzi.
+* Users would have no reason to manipulate the `metadata.version` on their own.
+
+The risk can be also at least partially mitigated through documentation.
```

Comment:
Would it make sense to raise a warning in Kafka CR when Strimzi detects manual change of `metadata.version` ?

Reply:
I'm not sure we can track it reliably. So I do not think we should have any special handling for it.